Move internal/x/exec to internal/exec

peter-edge opened the pull request.

AlexAPB commented on internal/exec/runner.go:
```diff
@@ -0,0 +1,922 @@
+
+type meta struct {
+	ProtoSets               []*file.ProtoSet
+	InDirModeSingleFilename string
```
can you comment this? It wasn't clear as to what this was indicating initially.

Might be worth separating this as two args: `bool` and `string`, but that's optional

peter-edge replied:
Separating leads to more bugs here potentially, think of this as an atomic thing, no inconsistent state where `dirMode == false && singleFilename == non-empty`. Commented.

AlexAPB commented on internal/exec/runner.go:
```diff
@@ -0,0 +1,922 @@
+	}
+}
+
+// TODO: this is copied in three places
```
We could put this in the `strs` directory, not a perfect spot, but better than repeating

peter-edge replied:
`strs` no, you wouldn't put this in `https://golang.org/pkg/strings`, which is what `strs` corresponds to. We can make a new package, filed https://github.com/uber/prototool/issues/133.

peter-edge commented on internal/exec/exec.go:
```diff
@@ -0,0 +1,131 @@
+// The args given are the args from the command line.
+// Each additional parameter generally refers to a command-specific flag.
+type Runner interface {
+	Init(args []string, uncomment bool) error
```
I know you like alphabetizing, but there's a general limit - the functions here are grouped by functionality more or less. For example, `BinaryToJSON` and `JSONToBinary` should be next to each other, including in the implementtion in `runner.go`, and `Files, Compile, Gen`, `Download, Clean`, etc. Are you sure you want to do this?

peter-edge commented on internal/exec/runner.go:
```diff
@@ -0,0 +1,922 @@
+var jsonMarshaler = &jsonpb.Marshaler{Indent: "  "}
+
+type runner struct {
+	configProvider   settings.ConfigProvider
```
They already are:

- `configProvider, protoSetProvdider` are created inside the constructor
- `workDirPath, input, output` are the parameters to the constructor, in order
- everything else are values for `RunnerOptions`, in the order they are declared in `exec.go`

Do you just want a newline in between `protoSetProvider, workDirPath` and `output, logger`?

peter-edge commented on internal/exec/runner.go:
```diff
@@ -0,0 +1,922 @@
+	if err != nil {
+		return err
+	}
+	if message == nil {
```
Actually turns out no, but I had to look through the code in `extract` to determine what I did - we return error if there was no message found. Added a comment in `extract` about the return values.

peter-edge commented on internal/exec/runner.go:
```diff
@@ -0,0 +1,922 @@
+					if _, ok := err.(*ExitError); !ok {
+						return err
+					}
+					retErr = err
```
I looked through the implementation, and the detail here, which is poorly abstracted, is that an `ExitError` in this context, contrary  to it's naming, is a non-system error where we had a diff on format, for example, and we don't want to exit until we process all the files. So if you pas `-l, -d`, or nothing, you want a non-zero exit code if there is a diff, but you don't want to stop until all files are processed. I refactored this a bit so it's a bit more clear, so that an error is actually an error, but it's still not great.

peter-edge commented on internal/exec/runner.go:
```diff
@@ -0,0 +1,922 @@
+		return newExitErrorf(255, "")
+	}
+	if !bytes.Equal(input, data) {
+		if overwrite {
```
I just added it to `Format`:

```go
	if (overwrite && diffMode) || (overwrite && lintMode) || (diffMode && lintMode) {
		return newExitError(255, "can only set one of overwrite, diff, lint")
	}
```

peter-edge commented on internal/exec/runner.go:
```diff
@@ -0,0 +1,922 @@
+		args = []string{"."}
+	}
+	if len(args) == 1 {
+		fileInfo, err := os.Stat(args[0])
```
Sure why not :-)

peter-edge commented on internal/exec/runner.go:
```diff
@@ -0,0 +1,922 @@
+		}
+		// TODO: allow symlinks?
+		if !fileInfo.Mode().IsRegular() {
+			return nil, fmt.Errorf("multiple arguments only allowed if all arguments are regular files and %s is not a regular file", args[0])
```
Yes, whoops. And comma instead of period, this is a returned error

peter-edge commented:
I addressed all comments, but some of them I disagreed with so can you take another look and make sure this is OK? On the file splitting, we can do that, I personally never cared, but let's leave that for later.
